File: shared/protocol/src/binary-frame.ts
/**
 * Binary-frame envelope helpers.
 *
 * Wire layout: `[1 byte type][2 bytes LE header_len][header bytes][payload]`
 *
 * Header bytes are proto-encoded (`AudioChunkHeader` or `ImageChunkHeader`).
 * Decoding is caller-driven per `frameType` — no polymorphic envelope needed.
 */

import { create, fromBinary, toBinary } from '@bufbuild/protobuf'
import {
  type AudioChunkHeader,
  AudioChunkHeaderSchema,
  type ImageChunkHeader,
  ImageChunkHeaderSchema,
  BinaryFrameType,
} from './generated/moumantai/v1/index.js'

const HEADER_LEN_BYTES = 2
const TYPE_BYTE_LEN = 1
const ENVELOPE_PREFIX_LEN = TYPE_BYTE_LEN + HEADER_LEN_BYTES // 3
const HEADER_MAX = 0xffff // uint16

export interface ParsedBinaryFrame {
  frameType: BinaryFrameType
  /** Proto-encoded header bytes; decode via `decodeAudioHeader` / `decodeImageHeader`. */
  headerBytes: Uint8Array
  payload: Uint8Array
}

/** Parse a binary frame's outer envelope. Returns null on malformed bytes. */
export function parseBinaryFrame(data: Uint8Array): ParsedBinaryFrame | null {
  if (data.length < ENVELOPE_PREFIX_LEN) return null

  const frameType = data[0] as BinaryFrameType
  const headerLen = data[1]! | (data[2]! << 8)
  if (data.length < ENVELOPE_PREFIX_LEN + headerLen) return null

  return {
    frameType,
    headerBytes: data.subarray(ENVELOPE_PREFIX_LEN, ENVELOPE_PREFIX_LEN + headerLen),
    payload: data.subarray(ENVELOPE_PREFIX_LEN + headerLen),
  }
}

/** Build a binary frame from a typed frame-type, raw header bytes, and payload. */
export function encodeBinaryFrame(
  frameType: BinaryFrameType,
  headerBytes: Uint8Array,
  payload: Uint8Array,
): Uint8Array {
  if (headerBytes.length > HEADER_MAX) {
    throw new Error(
      `Binary frame header exceeds uint16 max (${headerBytes.length} > ${HEADER_MAX})`,
    )
  }
  const frame = new Uint8Array(ENVELOPE_PREFIX_LEN + headerBytes.length + payload.length)
  frame[0] = frameType
  frame[1] = headerBytes.length & 0xff
  frame[2] = (headerBytes.length >> 8) & 0xff
  frame.set(headerBytes, ENVELOPE_PREFIX_LEN)
  frame.set(payload, ENVELOPE_PREFIX_LEN + headerBytes.length)
  return frame
}

// ---------------------------------------------------------------------------
// Type-safe convenience helpers
// ---------------------------------------------------------------------------

export function encodeAudioFrame(header: AudioChunkHeader, payload: Uint8Array): Uint8Array {
  return encodeBinaryFrame(BinaryFrameType.AUDIO, toBinary(AudioChunkHeaderSchema, header), payload)
}

export function decodeAudioHeader(headerBytes: Uint8Array): AudioChunkHeader {
  return fromBinary(AudioChunkHeaderSchema, headerBytes)
}

export function encodeImageFrame(header: ImageChunkHeader, payload: Uint8Array): Uint8Array {
  return encodeBinaryFrame(BinaryFrameType.IMAGE, toBinary(ImageChunkHeaderSchema, header), payload)
}

export function decodeImageHeader(headerBytes: Uint8Array): ImageChunkHeader {
  return fromBinary(ImageChunkHeaderSchema, headerBytes)
}

/** Convenience constructor for `AudioChunkHeader`. */
export function audioChunkHeader(fields: {
  scope?: string
  format: AudioChunkHeader['format']
  sampleRate: number
  final: boolean
  clientMsgId?: string
}): AudioChunkHeader {
  return create(AudioChunkHeaderSchema, {
    scope: fields.scope ?? '',
    format: fields.format,
    sampleRate: fields.sampleRate,
    final: fields.final,
    clientMsgId: fields.clientMsgId,
  })
}

File: shared/protocol/kotlin/com/moumantai/transport/BinaryFrame.kt
package com.moumantai.transport

import com.moumantai.protocol.v1.AudioChunkHeader
import com.moumantai.protocol.v1.AudioFormat
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Shared binary-frame envelope helpers for Android + Wear OS.
 *
 * Wire layout: `[1 byte type][2 bytes LE header_len][header bytes][payload]`.
 * The header bytes are proto-encoded `AudioChunkHeader` (chat.proto). Both
 * clients consume + produce the exact same shape.
 *
 * `Type.IMAGE` is retained as a defensive no-op route on the inbound side of
 * each client — no current code path emits IMAGE frames (image input travels
 * inline on `ChatInput.image_data`), but the constant guards the binary-frame
 * router against accidental mis-decode.
 *
 * The proto enum [com.moumantai.protocol.v1.BinaryFrameType] is `Int`-valued
 * via `WireEnum`; for the leading-byte channel tag we expose the same
 * values as [Byte] under [Type].
 */
object BinaryFrame {

    object Type {
        val AUDIO: Byte = com.moumantai.protocol.v1.BinaryFrameType.BINARY_FRAME_TYPE_AUDIO.value.toByte()
        val IMAGE: Byte = com.moumantai.protocol.v1.BinaryFrameType.BINARY_FRAME_TYPE_IMAGE.value.toByte()
    }

    /** Decoded view of a binary frame; [headerBytes] is proto-encoded. */
    data class Parsed(val type: Byte, val headerBytes: ByteArray, val payload: ByteArray)

    /** Parse the outer envelope. Returns null on undersized or oversized frames. */
    fun parse(frame: ByteArray): Parsed? {
        if (frame.size < 3) return null
        val headerLen = ByteBuffer.wrap(frame, 1, 2).order(ByteOrder.LITTLE_ENDIAN).short.toInt() and 0xFFFF
        if (frame.size < 3 + headerLen) return null
        return Parsed(
            type = frame[0],
            headerBytes = frame.copyOfRange(3, 3 + headerLen),
            payload = frame.copyOfRange(3 + headerLen, frame.size),
        )
    }

    /** Build a frame from a typed-enum tag, raw header bytes, and payload. */
    fun encode(type: Byte, headerBytes: ByteArray, payload: ByteArray): ByteArray {
        require(headerBytes.size <= 0xFFFF) { "Binary frame header exceeds uint16 max" }
        val frame = ByteBuffer.allocate(1 + 2 + headerBytes.size + payload.size).order(ByteOrder.LITTLE_ENDIAN)
        frame.put(type)
        frame.putShort(headerBytes.size.toShort())
        frame.put(headerBytes)
        frame.put(payload)
        return frame.array()
    }

    // ---- Type-safe convenience helpers ----

    fun encodeAudio(
        scope: String,
        audioData: ByteArray,
        format: String,
        sampleRate: Int,
        isFinal: Boolean,
        clientMsgId: String?,
    ): ByteArray {
        val header = AudioChunkHeader(
            scope = scope,
            format = audioFormatProto(format),
            sample_rate = sampleRate,
            final_ = isFinal,
            client_msg_id = clientMsgId,
        )
        return encode(Type.AUDIO, AudioChunkHeader.ADAPTER.encode(header), audioData)
    }

    /** Decode the audio-frame proto header; throws on malformed bytes. */
    fun decodeAudio(headerBytes: ByteArray): AudioChunkHeader = AudioChunkHeader.ADAPTER.decode(headerBytes)

    // ---- Format <-> string helpers ----
    //
    // Only PCM16 is currently produced or consumed; the AudioFormat enum
    // carries no other codecs.

    fun audioFormatProto(label: String): AudioFormat = when (label) {
        "pcm16" -> AudioFormat.AUDIO_FORMAT_PCM16
        else    -> AudioFormat.AUDIO_FORMAT_UNSPECIFIED
    }

    fun audioFormatLabel(value: AudioFormat): String? = when (value) {
        AudioFormat.AUDIO_FORMAT_PCM16 -> "pcm16"
        AudioFormat.AUDIO_FORMAT_UNSPECIFIED -> null
    }
}
